File: OOP/abstraction.py
from abc import ABC,abstractmethod
from math import pi

class Shape(ABC):
    @abstractmethod
    def area(self):
        pass
    @abstractmethod
    def perimeter(self):
        pass


class Circle(Shape):
    def __init__(self,radius) -> None:
        self.radius = radius

    def area(self):
        print(self.radius*self.radius*pi)

    def perimeter(self):
        print(self.radius*2*pi)


class Rectangle(Shape):
    def __init__(self,height,width) -> None:
        self.height = height
        self.width = width

    def area(self):
        print(self.height*self.width)

    def perimeter(self):
        print((self.height+self.width)*2)

rectangle = Rectangle(3,4)
circle = Circle(4)
rectangle.area()
rectangle.perimeter()
circle.area()
circle.perimeter()

File: OOP/abstraction.cpp
#include <iostream>
#include <string>
#define M_PI 3.141592654
using namespace std;

class Shape{
    public:
        virtual double area() = 0;
        virtual double perimeter() = 0;
};

class Circle:public Shape{
    private:
        double radius;
    public:
        Circle(double r) : radius(r){};

        double area() override{
            return radius*radius*M_PI;
        }

        double perimeter() override{
            return radius*2*M_PI;
        }
};

class Rectangle:public Shape{
    private:
        double height;
        double width;
    public:
        Rectangle(double h,double w) : height(h),width(w){};

        double area() override{
            return height*width;
        }

        double perimeter() override{
            return 2*(height+width);
        }
};